### PR TITLE
Add new GVFS Config verb

### PR DESCRIPTION
**Pull request by alameenshah**
#### Summary
- New ConfigVerb.
- Usage gvfs config <key> <value>
- Renamed GVFSConfig.cs to ServerGVFSConfig.cs.
- New LocalGVFSConfig.cs file. It writes config key value pairs into a config file.
- Updated ProductUpgrader.cs to use the new LocalGVFSConfig class to read ring config.
- Cleanup - renamed all occurances of gvfsConfig to serverGVFSConfig.

#### File Format
Config file is a simple dictionary file stored at path C:\ProgramData\GVFS\local_config.dat. 

#### Command usage and messaging:
```
PS M:\Work\gvfs\VFSForGit_Ameen> gvfs config upgrade.ring "Fast"
PS M:\Work\gvfs\VFSForGit_Ameen> gvfs config upgrade.ring
Fast
```

### CODE REVIEW

File: GVFS/GVFS.Common/LocalGVFSConfig.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        private bool TryGetConfig(string key, out string value, out string error)
+        {
+            if (!this.fileSystem.FileExists(this.configFile))
+            {
+                error = $"Error reading {key}. Config file({this.configFile}) does not exist.";
+                value = null;
+                return false;
+            }
+
+            GitProcess.Result result = GitProcess.GetFromFileConfig(this.gitPath, this.configFile, key);
```
Comment (sanoursa) [inline GVFS/GVFS.Common/LocalGVFSConfig.cs]: This is a definite downside to using git's format - we have to launch a process every time we want to read or write a value. Using our own file format would not have that issue.

File: GVFS/GVFS.Common/ProductUpgrader.cs [hunk 2]
```diff
@@ -261,14 +262,10 @@ public virtual bool TryLoadRingConfig(out string error)
                 }
                 else
                 {
-                    error = "Invalid upgrade ring `" + ringConfig + "` specified in Git config.";                    
+                    error = "Invalid upgrade ring `" + ringConfig + "` specified in Git config.";
```
Comment (sanoursa) [inline GVFS/GVFS.Common/ProductUpgrader.cs]: It's no longer in the "Git config". Please scrub any other messages that say this.

File: GVFS/GVFS.Common/ProductUpgrader.cs [hunk 3]
```diff
@@ -245,13 +245,14 @@ public bool TryCleanup(out string error)
         public virtual bool TryLoadRingConfig(out string error)
         {
             string gitPath = GVFSPlatform.Instance.GitInstallation.GetInstalledGitBinPath();
-            GitProcess.Result result = GitProcess.GetFromGlobalConfig(gitPath, GVFSConstants.GitConfig.UpgradeRing);
-            if (!result.HasErrors && !string.IsNullOrEmpty(result.Output))
+            LocalGVFSConfig localConfig = new LocalGVFSConfig(gitPath);
+
+            string ringConfig = null;
+            if (localConfig.TryGetValueForKey(GVFSConstants.GitConfig.UpgradeRing, out ringConfig, out error))
```
Comment (sanoursa) [inline GVFS/GVFS.Common/ProductUpgrader.cs]: Move the `UpgradeRing` constant out of the `GitConfig` class.

File: GVFS/GVFS/CommandLine/ConfigVerb.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
+﻿using CommandLine;
+using GVFS.Common;
+using System;
+
+namespace GVFS.CommandLine
+{
+    [Verb(ConfigVerbName, HelpText = "Set and Get GVFS config settings.")]
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/ConfigVerb.cs]: "Get" should be lowercase

File: GVFS/GVFS/CommandLine/ConfigVerb.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace GVFS.CommandLine
+{
+    [Verb(ConfigVerbName, HelpText = "Set and Get GVFS config settings.")]
+    public class ConfigVerb : GVFSVerb
+    {
+        private const string ConfigVerbName = "config";
+        private LocalGVFSConfig localConfig;
+
+        public override string EnlistmentRootPathParameter { get; set; }
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/ConfigVerb.cs]: Throw if this property is accessed (see `ServiceVerb` as an example). This command is not tied to any particular repo.

File: GVFS/GVFS/CommandLine/ConfigVerb.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        protected override string VerbName
+        {
+            get { return ConfigVerbName; }
+        }
+
+        public override void Execute()
+        {
+            string[] args = Environment.GetCommandLineArgs();
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/ConfigVerb.cs]: Please don't parse the command line args directly. Use our existing command line parsing patterns.

File: GVFS/GVFS/Program.cs [hunk 7]
```diff
@@ -27,7 +27,8 @@ public static void Main(string[] args)
                 typeof(ServiceVerb),
                 typeof(StatusVerb),
                 typeof(UnmountVerb),
-                typeof(UpgradeVerb)
+                typeof(UpgradeVerb),
+                typeof(ConfigVerb)
```
Comment (sanoursa) [inline GVFS/GVFS/Program.cs]: Leave a trailing comma

File: GVFS/GVFS/Program.cs [hunk 8]
```diff
@@ -89,6 +90,12 @@ public static void Main(string[] args)
                             upgrade.Execute();
                             Environment.Exit((int)ReturnCode.Success);
                         })
+                    .WithParsed<ConfigVerb>(
```
Comment (sanoursa) [inline GVFS/GVFS/Program.cs]: `ServiceVerb`, `UpgradeVerb`, and now `ConfigVerb` all follow the same pattern - they are tied to machine state, not a particular repo. Please make a subclass like `GVFSVerb.NonRepoVerb` (see `GVFSVerb.ForExistingEnlistment` as an example) so that we can just have one `.WithParsed<GVFSVerb.NonRepoVerb>` here, and also eliminate the boilerplate in the implementations of all of those verbs.

File: GVFS/GVFS/CommandLine/ConfigVerb.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            if (args.Length < 3 || args.Length > 4)
+            {
+                string usageString = string.Join(
+                    Environment.NewLine,
+                    "Error: wrong number of arguments.",
+                    "Usage: gvfs config <config> <value>");
```
Comment (sanoursa) [inline GVFS/GVFS/CommandLine/ConfigVerb.cs]: This usage message is incorrect, since value is optional and causes a different behavior. Please be more complete here.

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 10]
```diff
@@ -227,7 +227,7 @@ public static class UpgradeVerbMessages
             public const string GVFSUpgradeOptionalConfirm = "`gvfs upgrade [--confirm]`";
             public const string NoneRingConsoleAlert = "Upgrade ring set to \"None\". No upgrade check was performed.";
             public const string InvalidRingConsoleAlert = "Upgrade ring is not set. No upgrade check was performed.";
-            public const string SetUpgradeRingCommand = "To set or change upgrade ring, run `git config --global " + GitConfig.UpgradeRing + " [\"Fast\"|\"Slow\"|\"None\"] from a command prompt.";
+            public const string SetUpgradeRingCommand = "To set or change upgrade ring, run `gvfs config " + GitConfig.UpgradeRing + " [\"Fast\"|\"Slow\"|\"None\"] from an elevated command prompt.";
```
[the remark this answers is not in the document]
Comment (jrbriggs) [reply GVFS/GVFS.Common/GVFSConstants.cs]: It should not need to be elevated. Let's remove that.

File: GVFS/GVFS.Common/ServerGVFSConfig.cs [hunk 11]
```diff
@@ -5,7 +5,7 @@
 
 namespace GVFS.Common
 {
-    public class GVFSConfig
+    public class ServerGVFSConfig
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/ServerGVFSConfig.cs]: This would be easier to review if this rename was done in a separate commit. Can you tease those apart?

File: GVFS/GVFS/CommandLine/ConfigVerb.cs (hunk 4 above)
Comment (jrbriggs) [reply GVFS/GVFS/CommandLine/ConfigVerb.cs]: Also reword -- "Get and set GVFS options"

File: GVFS/GVFS.Common/GVFSConstants.cs [hunk 12]
```diff
@@ -33,7 +33,7 @@ public static class GitConfig
             public const string HooksPrefix = GitConfig.GVFSPrefix + "clone.default-";
             public const string GVFSTelemetryId = GitConfig.GVFSPrefix + "telemetry-id";
             public const string HooksExtension = ".hooks";
-            public const string UpgradeRing = GVFSPrefix + "upgrade-ring";
+            public const string UpgradeRing = "upgrade-ring";
```
Comment (jrbriggs) [inline GVFS/GVFS.Common/GVFSConstants.cs]: If we're lifting into our own config file, upgrade.ring so we could have other options in the upgrade section?

File: GVFS/GVFS/Program.cs (hunk 8 above)
Comment (sanoursa) [reply GVFS/GVFS/Program.cs]: For a consistent name, perhaps go for `GVFSVerb.ForNoEnlistment`. Or something else if you have a better suggestion.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Common/GVFSConstants.cs]: Removed.

File: GVFS/GVFS.Common/Git/GitProcess.cs [hunk 13]
```diff
@@ -210,6 +215,16 @@ public Result AddInLocalConfig(string settingName, string value)
                  value));
         }
 
+        public Result SetInFileConfig(string configFile, string settingName, string value, bool replaceAll = false)
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Common/Git/GitProcess.cs]: using FileBasedDictionary now.

File: GVFS/GVFS.Common/LocalGVFSConfig.cs (hunk 1 above)
Comment (alameenshah) [reply GVFS/GVFS.Common/LocalGVFSConfig.cs]: Ya... replaced with FileBasedDictionary

File: GVFS/GVFS.Common/ProductUpgrader.cs (hunk 3 above)
Comment (alameenshah) [reply GVFS/GVFS.Common/ProductUpgrader.cs]: Moved it to LocalGVFSConfig class

File: GVFS/GVFS/Program.cs (hunk 8 above)
Comment (alameenshah) [reply GVFS/GVFS/Program.cs]: Renamed to GVFSVerb.ForNoEnlistment

File: GVFS/GVFS.Common/ServerGVFSConfig.cs (hunk 11 above)
Comment (alameenshah) [reply GVFS/GVFS.Common/ServerGVFSConfig.cs]: Moved the rename of GVFSConfig to a separate commit.

File: GVFS/GVFS.Common/GVFSConstants.cs (hunk 12 above)
Comment (alameenshah) [reply GVFS/GVFS.Common/GVFSConstants.cs]: Sure. Renamed it.

**Comment by kewillford** [comment]
This PR is changing a bunch of places we are using "GVFS".  Does it make m sense to start using either "VFSForGit" or using something like "App" so that we don't continue using the GVFS naming?
